### PR TITLE
[RDY] Basic implementation of shell.core

### PR DESCRIPTION
**Pull request by Hjdskes**
Basic as in "enough to have a demo by Friday". Not everything is set in stone yet, but you can see the `TODO`s in the code for that.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -12,9 +12,19 @@
     <dependencyManagement>
         <dependencies>
             <dependency>
+                <groupId>com.google.inject</groupId>
+                <artifactId>guice</artifactId>
+                <version>4.0</version>
+            </dependency>
+            <dependency>
+                <groupId>com.google.inject.extensions</groupId>
+                <artifactId>guice-multibindings</artifactId>
```
Comment (Hjdskes) [inline pom.xml:21]: This dep needs to be explicitly inherited by modules that use it. Right now it's an indirect dependency through the spoofax dependency, which is why the build still works.

File: org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/commands/EvaluateCommand.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    /**
+     * Evaluates {@code args}.
+     * @param args The expression(s) to evaluate.
+     */
+    @Override
+    public void execute(String... args) {
+        try {
+            this.onSuccess.accept(new StyledText(common.toString(this.interp(args[0]))));
+        } catch (IOException | MetaborgException e) {
+            e.printStackTrace();
```
Comment (Hjdskes) [inline org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/commands/EvaluateCommand.java]: I think we need to switch to a logger: these stacktraces are printed to `System.err` and therefore they appear in the REPL UI.

File: org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/commands/EvaluateCommand.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+     */
+    public IStrategoTerm interp(ISpoofaxAnalyzeUnit analyzeUnit) throws MetaborgException {
+        final FacetContribution<StrategoRuntimeFacet> runContrib = this.context.language()
+                .facetContribution(StrategoRuntimeFacet.class);
+        if (runContrib == null) {
+            throw new MetaborgException("Cannot retrieve the runtime facet");
+        }
+        final HybridInterpreter interpreter = runtimeService.runtime(
+                runContrib.contributor, this.context);
+        return common.invoke(interpreter, analyzeUnit.ast(), "runstrat");
```
Comment (Hjdskes) [inline org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/commands/EvaluateCommand.java:87]: Needs to be removed in a later commit / separate PR.

File: org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/client/ReplModule.java [hunk 4]
```diff
@@ -25,6 +42,13 @@
     protected void configureCommands() {
         commandBinder = MapBinder.newMapBinder(binder(), String.class, IReplCommand.class);
         commandBinder.addBinding("exit").to(Repl.ExitCommand.class).in(Singleton.class);
+        commandBinder.addBinding("help").to(HelpCommand.class).in(Singleton.class);
+        commandBinder.addBinding("parse").to(ParseCommand.class).in(Singleton.class);
+        commandBinder.addBinding("analyze").to(AnalyzeCommand.class).in(Singleton.class);
+        // commandBinder.addBinding("eval").to(EvaluateCommand.class).in(Singleton.class);
+        // FIXME: DIRTY HACK
```
Comment (Hjdskes) [inline org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/client/ReplModule.java]: Can this be resolved before this PR is merged?

**Comment by Hjdskes** [comment]
I think we should merge `ReplModule` and `CoreModule`.

Does the implicit requirement still exist that, say, `EvalCommand.execute` should not use `AnalyzeCommand.execute`? If so, this needs to be documented.

Can the "hardcoded" analysis step be skipped in this PR, or would that be too much work?

Lastly, are you planning on writing more tests before this gets merged?

Unresolved but collapsed comment: https://github.com/spoofax-shell/spoofax-shell/pull/3#discussion-diff-63225034

File: pom.xml (hunk 1 above)
Comment (Balletie) [inline pom.xml:21]: Why do we use 4.0? Spoofax uses 3.0. I am not sure if a mixed installation is a wise idea.

Comment (Hjdskes) [inline pom.xml:21]: Because 4.0 provides error messages when something fails. It was you that found this :P

Comment (Balletie) [inline pom.xml:21]: yea but I didn't commit it, I don't think a mixed installation would work well.

Comment (Hjdskes) [inline pom.xml:21]: Oke, let's change this to 3.0 then.

Comment (Hjdskes) [inline pom.xml:21]: Keeping at 4.0: https://github.com/google/guice/wiki/Guice40#user-content-changes-since-guice-30

**Comment by Hjdskes** [comment]
PR is incomplete (lacking tests and unfinished comments), but is merged anyway such that the sprint tag has a workable REPL.

**Pull request merged**